File: src/test/java/test/java/BenchTest.java
package test.java;

import me.saro.dat.DatUtils;
import me.saro.dat.Unixtime;
import me.saro.dat.crypto.DatCryptoAlgorithm;
import me.saro.dat.dat.DatCertificate;
import me.saro.dat.dat.DatManager;
import me.saro.dat.dat.Payload;
import me.saro.dat.signature.DatSignatureAlgorithm;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.stream.IntStream;

public class BenchTest {

    private IntStream stream(boolean multiThread, int loop) {
        IntStream stream = IntStream.range(0, loop);
        return multiThread ? stream.parallel() : stream.sequential();
    }

    public void loop(boolean multiThread, int loop, List<DatCertificate> certificates, String plain, String secure) {

        System.out.println("\n" + (multiThread ? "Multi-Thread " : "Single-Thread "));

        for (DatCertificate certificate : certificates) {
            String pre = certificate.getSignature$dat().algorithm() + " " +
                    certificate.getCrypto$dat().algorithm() + " ";

            long time = System.currentTimeMillis();
            List<String> dats = stream(multiThread, loop)
                    .mapToObj(i -> DatManager.issue(certificate, plain, secure))
                    .toList();
            System.out.println(pre + "Issue * " + dats.size() + " : " + (System.currentTimeMillis() - time) + "ms");

            time = System.currentTimeMillis();
            String dat = dats.get(0);
            List<Payload> payloads = stream(multiThread, loop)
                    .mapToObj(i -> DatManager.parse(certificate, dat))
                    .toList();

            System.out.println(pre + "Parse * " + payloads.size() + " : " + (System.currentTimeMillis() - time) + "ms");

            assert plain.equals(payloads.get(0).getPlain());
            assert secure.equals(payloads.get(0).getSecure());
        }
    }
    
    //@Test
    public void test() {
        int loop = 10000;
        long now = Unixtime.now();
        String plain = DatUtils.generateRandomBase62(100);
        String secure = DatUtils.generateRandomBase62(100);

        System.out.println("Plain : " + plain);
        System.out.println("Secure : " + secure);

        List<DatCertificate> certificates = DatSignatureAlgorithm.getEntries().stream().flatMap(sa ->
            DatCryptoAlgorithm.getEntries().stream().map(ca ->
                    DatCertificate.generate(0, now - 10, 200, 100, sa, ca)
            )
        ).toList();

        loop(true, loop, certificates, plain, secure);
        loop(false, loop, certificates, plain, secure);
    }
}

File: src/test/kotlin/test/kt/BenchTest.kt
package test.kt

import me.saro.dat.DatUtils.Companion.generateRandomBase62
import me.saro.dat.Unixtime.Companion.now
import me.saro.dat.crypto.DatCryptoAlgorithm
import me.saro.dat.dat.DatCertificate
import me.saro.dat.dat.DatManager.Companion.issue
import me.saro.dat.dat.DatManager.Companion.parse
import me.saro.dat.dat.Payload
import me.saro.dat.signature.DatSignatureAlgorithm
import org.junit.jupiter.api.Test
import java.util.function.IntFunction
import java.util.stream.IntStream

class BenchTest {
    private fun stream(multiThread: Boolean, loop: Int): IntStream {
        val stream = IntStream.range(0, loop)
        return if (multiThread) stream.parallel() else stream.sequential()
    }

    fun loop(
        multiThread: Boolean,
        loop: Int,
        certificates: MutableList<DatCertificate>,
        plain: String,
        secure: String
    ) {
        println("\n" + (if (multiThread) "Multi-Thread " else "Single-Thread "))

        for (certificate in certificates) {
            val pre = certificate.signatureKey.algorithm().toString() + " " +
                    certificate.cryptoKey.algorithm() + " "

            var time = System.currentTimeMillis()
            val dats = stream(multiThread, loop)
                .mapToObj<String>(IntFunction { i: Int -> issue(certificate, plain, secure) })
                .toList()
            println(pre + "Issue * " + dats.size + " : " + (System.currentTimeMillis() - time) + "ms")

            time = System.currentTimeMillis()
            val dat = dats.get(0)
            val payloads = stream(multiThread, loop)
                .mapToObj<Payload>(IntFunction { i: Int -> parse(certificate, dat) })
                .toList()

            println(pre + "Parse * " + payloads.size + " : " + (System.currentTimeMillis() - time) + "ms")

            assert(plain == payloads.get(0)!!.plain)
            assert(secure == payloads.get(0)!!.secure)
        }
    }

    //@Test
    fun test() {
        val loop = 10000
        val now = now()
        val plain = generateRandomBase62(100)
        val secure = generateRandomBase62(100)

        println("Plain : " + plain)
        println("Secure : " + secure)

        val certificates: MutableList<DatCertificate> = DatSignatureAlgorithm.entries.stream()
            .flatMap<DatCertificate> { sa: DatSignatureAlgorithm? ->
                DatCryptoAlgorithm.entries.stream().map<DatCertificate> { ca: DatCryptoAlgorithm? ->
                    DatCertificate.generate(
                        0,
                        sa!!,
                        ca!!,
                        now - 10,
                        now + 600,
                        60
                    )
                }
            }.toList()

        loop(true, loop, certificates, plain, secure)
        loop(false, loop, certificates, plain, secure)
    }
}
